### DISCUSSION TITLE
HP GK100F keyboard no keystrokes detected

### DISCUSSION THREAD

**Discussion by jeroentaverne** [hathach/tinyusb#1636] [General]
Hardware: Raspberry Pi Pico acting as USB host
Software: TinyUSB example cdc_msc_hid
Log: see attached file

This keyboard somehow shows up as a keyboard and mouse, but doesn't have any mouse/touchpad/trackball capabilities. Keyboard protocol shows up as instance 0, mouse protocol as instance 1, but HID_ITF_PROTOCOL_NONE isn't reported by this keyboard. After the logging output stops, the RGB lightning is not enabled and no keystrokes can be received. After I change the code by executing tuh_hid_set_report(dev_addr, instance, 0, HID_REPORT_TYPE_OUTPUT, (void*)value, 1) when the keyboard protocol is detected, the RGB lighting of the keyboard is enabled, and the keystrokes are actually received.

Is there an USB HID incompatibility in this keyboard, or should TinyUSB be aware of this issue to get these kind of keyboards working out of the box without example modification? What would be the correct location in the example code to execute tuh_hid_set_report? This because after my modification the mouse protocol isn't detected anymore, so any real keyboard/mouse hardware combo detection would fail.

Thanks!

[HP_GK100F.txt](https://github.com/hathach/tinyusb/files/9547441/HP_GK100F.txt)

**Comment by hathach** [comment]
thanks for the report, HID is a bit open/loosed in term of specs, therefore getting it work with most/all device will take some test and trial. Since I don't have this keyboard, would you provide a bit more info, in the `tuh_hid_set_report()`
- the instance = 0 for keyboard right 
- what is your `value` contents, is it simply 0x00 ?

the txt file does not include the tuh_hid_set_report(), would you also update txt to also include that call.

Anyway, I could already guess the reason for this keyboard. The tuh_hid_set_report() is for setting the LED (capslock, numlock) which typically set by PC host after enumeration. This keyboard implementation is probably waiting for this specific set report before enabling its function, which is out of HID specs, but again, this is done often enough on pc host. I will look up the HID specs to see if this is safe enough to be included as part of the enumeration for every keyboard.

**Reply by jeroentaverne** [reply]
This is the code I added. It will make the keystrokes work, but the "mouse" instance is not reported anymore. I will send you a logfile as well.

  printf("HID Interface Protocol = %s\r\n", protocol_str[itf_protocol]);
  
+  if (itf_protocol == HID_ITF_PROTOCOL_KEYBOARD )
+  {
+  	static uint8_t value = 0;
+  	printf("Executing tuh_hid_set_report\n");
+	tuh_hid_set_report(dev_addr, instance, 0, HID_REPORT_TYPE_OUTPUT, (void*)&value, 1);
+  }

  // By default host stack will use activate boot protocol on supported interface.

**Reply by hathach** [reply]
yup, please upload an log with modified code as well.

**Reply by jeroentaverne** [reply]
Perhaps it's also an idea to turn on the NUMLOCK indicator by default like almost all PCs do.
See attached logfile.

[HP_GK100F_with_set_report.txt](https://github.com/hathach/tinyusb/files/9555261/HP_GK100F_with_set_report.txt)

**Reply by hathach** [reply]
thanks for update log, as suspected, this keyboard seems to wait for set keyboard report before running its firmware. NUMLOCK indicator won't get set though, it is up to application. For the failed enumeration of mouse instance, it is due to control transfer is busy when setting report, I will update the enumeration to re-try instead of giving up. Will make an PR when having time.

**Reply by jeroentaverne** [reply]
Thanks a lot. I am controlling the capslock in my application so that's why I found out that the keyboard worked with my code, but not with the unmodified example code. Perhaps it's an idea to add caplock key detection and capslock LED control in the HID example.

**Reply by jeroentaverne** [reply]
Hi, has this ever been "fixed"?
